### ISSUE TITLE
ModelCheckpoint every validation cycle instead of epoch?

### ISSUE THREAD

**Issue by dvirginz** [PyTorchLightning/pytorch-lightning#3485]
Does `ModelCheckpoint` support checkpointing every validation?
In case of big sequential data loaders (without shuffling) one will want to save the model each `val_check_interval` instead of each end of epoch (As it is a sequential data loader, I cannot use the `limit_train_batches` as it will be as overfitting on a subsample of the dataset).
Is there an easy way to do so besides wrapping ModelCheckpoint with my own callback class?

**Comment by shade22413** [comment]
@dvirginz Did you found a solution to your problem? I'm facing a similar one.

**Comment by dvirginz** [comment]
By inspecting the code I am now sure it runs on validation end, thus it does what we want it to do(they need to fix the docs). 
Maybe someone from pl will validate me and I'll submit a pr for the docs.

**Comment by matthid** [comment]
Sive I've run into the same problem and found this thread: Currently the code doesn't allow if you want to save the top k checkpoints on every validation step... Best you can do is set `every_n_epochs` to `2` then it will work every 2nd epoch.

https://github.com/Lightning-AI/lightning/blob/54e8095a78eafad3f4e16b799198b7bce8d327d2/src/lightning/pytorch/callbacks/model_checkpoint.py#L310-L316C68

However you can overwrite the function:
```python
        checkpoint_callback = ModelCheckpoint(
            save_top_k=5,
            monitor="val_loss",
            every_n_train_steps=0,
            save_on_train_epoch_end=False, # Run at end of validation.
            mode="min",
            dirpath=checkpoint_dir,
            filename=experiment_name + "-{epoch:02d}-{val_loss:.2f}",
        )

        def on_validation_end(self:ModelCheckpoint, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
            """Save a checkpoint at the end of the validation stage."""
            if not self._should_skip_saving_checkpoint(trainer) and not self._should_save_on_train_epoch_end(trainer):
                monitor_candidates = self._monitor_candidates(trainer)
                self._save_topk_checkpoint(trainer, monitor_candidates)
                self._save_last_checkpoint(trainer, monitor_candidates)

        checkpoint_callback.on_validation_end = types.MethodType(on_validation_end, checkpoint_callback)
```

This will probably break in future versions.
